File: bg/okgame/src/renderer/puzzle/Block.ts
import { EventEmitter } from 'eventemitter3';
import { Color, Colors } from './Color';
import { BlockType, BlockTypes } from './BlockType';
import type { Piece } from './Piece';
import type { Grid } from './Grid';

export enum AnimationState {
  NORMAL = 0,
  DROPPING = 1,
  TOUCHING_BOTTOM = 2,
  SET_AT_BOTTOM = 3,
  FLASHING = 4,
  REMOVING = 5,
  PRESSURE = 6,
}

export interface BlockEvents {
  stateChange: (state: AnimationState) => void;
  colorChange: (color: Color) => void;
  remove: () => void;
}

export interface BlockConfig {
  blockType?: BlockType;
  color?: Color;
  xInPiece?: number;
  yInPiece?: number;
}

export class Block extends EventEmitter<BlockEvents> {
  private _blockType: BlockType;
  private _color: Color;

  xInPiece: number = 0;
  yInPiece: number = 0;
  xGrid: number = -1;
  yGrid: number = -1;

  piece: Piece | null = null;
  grid: Grid | null = null;

  animationState: AnimationState = AnimationState.NORMAL;
  animationFrame: number = 0;
  animationFrameTicks: number = 0;

  setInGrid: boolean = false;
  locking: boolean = false;
  lockingAnimationFrame: number = 0;

  fadingOut: boolean = false;
  disappearingAlpha: number = 1.0;

  flashingToBeRemoved: boolean = false;

  connectedUp: boolean = false;
  connectedDown: boolean = false;
  connectedLeft: boolean = false;
  connectedRight: boolean = false;

  counterCount: number = -1;

  ticksSinceLastMovement: number = 0;

  constructor(config?: BlockConfig) {
    super();
    this._blockType = config?.blockType ?? BlockTypes.NORMAL;
    this._color = config?.color ?? this._blockType.getRandomColor();
    this.xInPiece = config?.xInPiece ?? 0;
    this.yInPiece = config?.yInPiece ?? 0;

    if (this._blockType.counterType !== 0) {
      this.counterCount = this._blockType.counterStartAmount;
    }
  }

  get blockType(): BlockType {
    return this._blockType;
  }

  set blockType(value: BlockType) {
    this._blockType = value;
  }

  get color(): Color {
    return this._color;
  }

  set color(value: Color) {
    if (!this._color.equals(value)) {
      this._color = value;
      this.emit('colorChange', value);
    }
  }

  setRandomColor(): void {
    this.color = this._blockType.getRandomColor();
  }

  setState(state: AnimationState): void {
    if (this.animationState !== state) {
      this.animationState = state;
      this.animationFrame = 0;
      this.animationFrameTicks = 0;
      this.emit('stateChange', state);
    }
  }

  resetConnections(): void {
    this.connectedUp = false;
    this.connectedDown = false;
    this.connectedLeft = false;
    this.connectedRight = false;
  }

  setGridPosition(x: number, y: number): void {
    this.xGrid = x;
    this.yGrid = y;
  }

  setPieceOffset(x: number, y: number): void {
    this.xInPiece = x;
    this.yInPiece = y;
  }

  matchesColor(other: Block): boolean {
    return this._color.equals(other._color);
  }

  matchesType(other: Block): boolean {
    return this._blockType.uuid === other._blockType.uuid;
  }

  startFadeOut(): void {
    this.fadingOut = true;
    this.disappearingAlpha = 1.0;
  }

  updateFadeOut(dt: number, fadeSpeed: number = 1.0): boolean {
    if (!this.fadingOut) return false;

    this.disappearingAlpha -= dt * fadeSpeed;
    if (this.disappearingAlpha <= 0) {
      this.disappearingAlpha = 0;
      this.emit('remove');
      return true;
    }
    return false;
  }

  startLocking(): void {
    this.locking = true;
    this.lockingAnimationFrame = 0;
  }

  isLocking(): boolean {
    return this.locking;
  }

  finishLocking(): void {
    this.locking = false;
    this.setInGrid = true;
    this.setState(AnimationState.SET_AT_BOTTOM);
  }

  clone(): Block {
    const block = new Block({
      blockType: this._blockType,
      color: this._color,
      xInPiece: this.xInPiece,
      yInPiece: this.yInPiece,
    });
    block.xGrid = this.xGrid;
    block.yGrid = this.yGrid;
    block.animationState = this.animationState;
    block.setInGrid = this.setInGrid;
    block.counterCount = this.counterCount;
    return block;
  }
}

File: bg/bobsgameonlinejava/src/main/java/com/bobsgame/puzzle/Block.java
package com.bobsgame.puzzle;

import com.bobsgame.shared.BobColor;
import com.bobsgame.client.GLUtils;
import java.util.ArrayList;

public class Block {
    public GameLogic game = null;
    public Grid grid = null;

    public Piece piece = null;
    public ArrayList<Block> connectedBlocksByPiece = new ArrayList<>();
    public ArrayList<Block> connectedBlocksByColor = new ArrayList<>();

    public int xInPiece = 0;
    public int yInPiece = 0;

    public int xGrid = -1;
    public int yGrid = -1;

    public BlockType blockType = null;

    private BobColor color = null;

    public float effectAlphaFrom = 0.5f;
    public float effectAlphaTo = 0.8f;
    public long effectFadeTicksPerPhase = 1000;
    private float effectAlpha = 0.5f;
    private long effectFadeTicks = 0;
    private boolean effectFadeInOutToggle = false;

    public float colorFlashFrom = 0.0f;
    public float colorFlashTo = 1.0f;
    public long colorFlashTicksPerPhase = 100;
    private float colorFlash = 0.0f;
    private long colorFlashTicks = 0;
    private boolean colorFlashInOutToggle = false;

    public boolean overrideAnySpecialBehavior = false;

    public int interpolateSwappingWithX = 0;
    public int interpolateSwappingWithY = 0;
    public long swapTicks = 0;
    public boolean flashingToBeRemoved = false;
    public boolean flashingToBeRemovedLightDarkToggle = false;

    public boolean setInGrid = false;
    public boolean locking = false;
    public int lockingAnimationFrame = 0;
    public long lockAnimationFrameTicks = 0;

    public boolean fadingOut = false;
    public float disappearingAlpha = 1.0f;

    public float lastScreenX = -1;
    public float lastScreenY = -1;

    public long ticksSinceLastMovement = 0;

    public boolean slamming = false;
    private long ticksSinceSlam = 0;
    public float slamX = 0;
    public float slamY = 0;

    public AnimationState animationState = AnimationState.NORMAL;
    public int animationFrame = 0;
    public long animationFrameTicks = 0;
    public int animationFrameSpeed = 100;

    public int counterCount = -2;

    public boolean didFlashingColoredDiamond = false;
    public boolean ateBlocks = false;
    public int direction = -1;
    public long directionChangeTicksCounter = 0;

    public static final int UP = 0;
    public static final int LEFT = 1;
    public static final int DOWN = 2;
    public static final int RIGHT = 3;

    private int customInterpolationTicks = -1;

    public boolean popping = false;
    public boolean panic = false;

    public boolean connectedUp = false;
    public boolean connectedDown = false;
    public boolean connectedLeft = false;
    public boolean connectedRight = false;

    public boolean connectedUpRight = false;
    public boolean connectedDownRight = false;
    public boolean connectedUpLeft = false;
    public boolean connectedDownLeft = false;

    public Block() {}

    public Block(GameLogic game, Grid grid, Piece piece, BlockType blockType) {
        this.game = game;
        this.grid = grid;
        this.piece = piece;
        this.blockType = blockType;
    }

    public void update() {
        effectFadeTicks += game.ticks();
        ticksSinceLastMovement += game.ticks();

        if (slamming) {
            ticksSinceSlam += game.ticks();
            if (ticksSinceSlam >= 100) {
                slamming = false;
            }
        }

        if (locking) {
            lockAnimationFrameTicks += game.ticks();
            if (lockAnimationFrameTicks > 20) {
                lockAnimationFrameTicks = 0;
                if (lockingAnimationFrame < 8) {
                    lockingAnimationFrame++;
                } else {
                    lockingAnimationFrame = 0;
                    locking = false;
                }
            }
        }

        if (fadingOut) {
            if (disappearingAlpha > 0.0f) {
                disappearingAlpha -= (float)game.ticks() * 0.005f;
            }
            if (disappearingAlpha < 0.0f) disappearingAlpha = 0.0f;

            if (disappearingAlpha == 0.0f) {
                fadingOut = false;
                if(game.fadingOutBlocks.contains(this)) game.fadingOutBlocks.remove(this);
            }
        }
    }

    public void setXYOffsetInPiece(int x, int y) {
        this.xInPiece = x;
        this.yInPiece = y;
    }

    public void breakConnectionsInPiece() {
        if(connectedBlocksByPiece != null) connectedBlocksByPiece.clear();
        if(connectedBlocksByColor != null) connectedBlocksByColor.clear();
    }

    public float getScreenX() {
        if (grid == null) return 0;
        return grid.getXInFBO() + xGrid * grid.cellW();
    }

    public float getScreenY() {
        if (grid == null) return 0;
        return grid.getYInFBO() + yGrid * grid.cellH() + (grid.scrollPlayingFieldY / grid.scrollBlockIncrement) * grid.cellH();
    }

    public BobColor getColor() {
        return color;
    }

    public void setColor(BobColor color) {
        this.color = color;
    }

    public void setRandomBlockTypeColor() {
        if (blockType != null && !blockType.colors.isEmpty()) {
            int index = game.getRandomIntLessThan(blockType.colors.size(), "Block.setRandomColor");
            this.color = blockType.colors.get(index);
        }
    }

    public void render(float screenX, float screenY, float a, float scale, boolean interpolate, boolean ghost) {
        float w = grid.cellW() * scale;
        float h = grid.cellH() * scale;

        BobColor renderColor = color;
        if (renderColor == null) renderColor = BobColor.gray;

        if (slamming && ticksSinceSlam < 100) {
             float xDiff = screenX - slamX;
             float yDiff = screenY - slamY;
             screenX = slamX;
             screenY = slamY;
             // Stretch
             w += xDiff;
             h += yDiff;
        }

        if (interpolate && ticksSinceLastMovement < 100 && lastScreenX != -1) {
             // Simple lerp
             float t = (float)ticksSinceLastMovement / 100.0f;
             if (t > 1.0f) t = 1.0f;
             screenX = lastScreenX + (screenX - lastScreenX) * t;
             screenY = lastScreenY + (screenY - lastScreenY) * t;
        } else {
             lastScreenX = screenX;
             lastScreenY = screenY;
        }

        // Draw colored rectangle for now
        GLUtils.drawFilledRectXYWH(screenX, screenY, w, h, renderColor.rf(), renderColor.gf(), renderColor.bf(), a);

        // Outline
        GLUtils.drawBox(screenX, screenX+w, screenY, screenY+h, 255, 255, 255);
    }

    public void renderDisappearing() {
        render(getScreenX(), getScreenY(), disappearingAlpha, 1.0f + (2.0f - (disappearingAlpha * 2.0f)), true, false);
    }

    public void renderOutlines(float screenX, float screenY, float s) {
        // TODO
    }
}
